### DISCUSSION TITLE
run_async, MessageHandler

### DISCUSSION THREAD

**Discussion by ciromanfredi** [python-telegram-bot/python-telegram-bot#2354] [Q&A]
Hi, I would like to know how to solve the following problem. I have a message manager that only activates for photos that arrive at the bot and performs some operations. When run_async = False then I don't have any kind of problem and even if I send multiple photos together everything works correctly, but when more photos arrive at the same time (and run_async = True) the operations are performed only for the first photo . I would like to know how I can handle this.

**Comment by Bibo-Joshi** [comment]
Hi. Tbh, that's not a lot of information to work with. Please

* keep in mind that sending photo albums equals sending multiple photos one by one. albums are just multiple messages nicely displayed
* read the ask-right page in the wiki ;)

**Reply by ciromanfredi** [reply]
Hi, thank you for your reply and I apologize if I did not ask a question correctly. I already understood that if you send multiple photos to the bot, it will treat all the photos sent one by one. And I verify this in a "synchronous" callback, when I go to make asynchronous callbacks, run_async = True) only the first photo is managed and not all the others sent. Also I set the workers to 4. Please, can you show me a code example of receiving photos but with asynchronous callback? i suspect it is wrong to recover photos in callback method with:
def inphoto(update:Update,context:CallbackContext):
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;photo_file = update.message.photo [-1]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;...............

**Reply by Bibo-Joshi** [reply]
Can you please provide a minimal working example that does what you want with `run_async=False` and fails with `run_async=True`? Take `echobot.py` from the examples folder as basis. Then we can have a better look at what's going wrong

**Reply by ciromanfredi** [reply]
I noticed that the moment I add the handler directly, run_async = True works fine. While in my case I am using a conversationHandler and the internal handlers of the conversationHandler are run_async = True but this gives me problems ... is it possible? I am attaching an example.

NOT WORK!!!!
def start(...):
&nbsp;&nbsp;&nbsp;&nbsp;...
&nbsp;&nbsp;&nbsp;&nbsp;return BASE

def inphoto(...):
&nbsp;&nbsp;&nbsp;&nbsp;...
&nbsp;&nbsp;&nbsp;&nbsp;...
&nbsp;&nbsp;&nbsp;&nbsp;return BASE

conv_handler = ConversationHandler(
entry_points=[CommandHandler('start', start)],
&nbsp;&nbsp;&nbsp;&nbsp;states={
&nbsp;&nbsp;&nbsp;&nbsp;    BASE:[MessageHandler(Filters.photo,inphoto,run_async=True)]
&nbsp;&nbsp;&nbsp;&nbsp;},
&nbsp;&nbsp;&nbsp;&nbsp;fallbacks=[CommandHandler('stop', stop)],
&nbsp;&nbsp;&nbsp;&nbsp;)

dispatcher.add_handler(conv_handler)
###################################################################################

WORK
dispatcher.add_handler(MessageHandler(Filters.photo,inphoto,run_async=True))

i want fix first case..

**Reply by Bibo-Joshi** [reply]
please read this https://telegra.ph/Minimal-Working-Example-for-PTB-07-18

**Reply by ciromanfredi** [reply]
Hi... the code is here... thank you for taking the time to solve the problem....
https://pastebin.com/5z65jiTE

**Reply by Bibo-Joshi** [reply]
All right, then the "problem" is that you use the `run_async` handler inside a `ConversationHandler`. When the first photo comes in, the CH hands the update over to a worker that runs your callback. Now the CH immediately picks up the next photo, but can only process it if the worker is done with the first picture. As that's not yet the case, CH discards the photo. with  `run_async=False`, CH would wait for the worker to finish before picking up the next photo.

Note that you can 

a) make your `MassageHandler` run with `run_async=False` and still make the revelant bot methods calls in `inphoto` run asynchronously (see the `Performance Optimization` wiki page for details
b) use the `WAITING` state of CH to specify what should happen if the CH receives an update while waiting for the worker to finish (you could i.e. tell the user to try again in a few seconds)
